### ISSUE TITLE
Graph NG

### ISSUE THREAD

**Issue by dprokop** [grafana/grafana#27848]
This issue tracks work to be continued after https://github.com/grafana/grafana/pull/27572:

- [x] Data retrieval in plugins (access to data frame and correct field within it)
- [x] Hiding series makes panel to break  https://github.com/grafana/grafana/pull/28290
- [x] update uPlot to 1.2.x and update x-axis time formatting (ref https://github.com/leeoniya/uPlot/issues/303) (https://github.com/grafana/grafana/pull/28569) @dprokop 
- [x] switch to native `ms` mode in uPlot 1.4.5 (#29445)
- [x] Refactor plot initialization logic to use https://github.com/grafana/grafana/pull/29006 @dprokop 
- [ ] Bring back GraphNG and hooks tests
- [x] Bars, Staircase, Smooth, points-only rendering modes (https://github.com/grafana/grafana/pull/29359)
- [x] Explicit axis sizing for graph alignment (https://github.com/grafana/grafana/pull/29553)
- [x] Enable scale distribution config: 🏃 @dprokop https://github.com/grafana/grafana/pull/29684
- [x] Match hover point styles with Flot (https://github.com/grafana/grafana/pull/29866) @leeoniya 
- [x] Support for null gaps in smooth and staircase path builders (https://github.com/grafana/grafana/pull/29593) @leeoniya 
- [x] Band charts (fillDownTo) (https://github.com/grafana/grafana/pull/30268) @ryantxu
- [x] Series config - null as 0/connected (https://github.com/grafana/grafana/pull/28118, https://github.com/grafana/grafana/pull/29512)
- [ ] Annotations
  - [ ] Display annotations on the visualization
    - [x] Marks (#27972) 
    - [ ] Regions
  - [ ] Editor. Currently blocked by the lack of possibility to correctly attach mouse events to uPlot, but there is an open [proposal](https://github.com/leeoniya/uPlot/issues/298) that will enable this
    - [ ] https://github.com/grafana/grafana/pull/30419
- [ ] Legend plugin
  - [x] MVP(2e71401)
  - [x] Legend sizing
  - [x] Toggling series (#29575)
  - [x] Table legend stats (https://github.com/grafana/grafana/pull/30251)
    - [ ] Sorting
  - [x] Series color control via legend icon (currently series color is configured via field settings) (#30256)
- [x] Dark theme colors adjustments (text, axis grid etc)
- [x] Dark theme zoom region not visible 
- [x] Data links support via context menu
- [x] Abstraction over events rendered on top of the plot area to support annotations and exemplars in unified way (https://github.com/grafana/grafana/pull/28071)
- [x] Replace graph2 alpha panel with graph3
- [x] Sparkline @ryantxu #29632 
- [ ] Annotation & exemplar markers now use the same tooltip, which is pretty much a copy of each other. We need to implement a more generic component to handle rendering Markers with a tooltip
- [x] Streaming - update graph data only when relative time range is applied
- [ ] https://github.com/grafana/grafana/pull/27572#discussion_r488382540
- [ ] Options and field config - organize options for better UX
- [ ] Show/hide grid
- [x] Time zone support on x-axis
- [x] Refactor contexts to a single PlotContext to simplify render path
- [x] Graph Panel layout API
- [x] Color picker standard field config editor (https://github.com/grafana/grafana/pull/28039)
- [x] Revert MutableDataFrame interface change https://github.com/grafana/grafana/pull/27572/files#diff-a50d3269cb61b09314ba1d22ba2fab39
- [x] Detect real vs artificial gaps due to data join & alignment (https://github.com/grafana/grafana/pull/29250)
- [ ] Explicit & auto x/y padding sizing
- [ ] Clean field config properties when switching visualizations https://github.com/grafana/grafana/issues/29895

Toggling of: (#29575)
- [x] graph for series can be hidden.
- [x] tooltip for series can be hidden.
- [x] legend for series can be hidden.
- [x] implement matcher for "all except" (or use regex).
- [x] implement proper multi-radio config editor.
- [x] add tests
- [x] make it being "auto" configured when interacting with the graph.

uPlot issues
- [x] uPlot issue with x-axis ticks rendered outside graph https://github.com/leeoniya/uPlot/issues/362

Backlog:
- [ ] https://github.com/grafana/grafana/issues/28528
- [ ] **Error thrown on main gdev Graph NG dashboard**
  ![image](https://user-images.githubusercontent.com/2376619/104745269-326cf880-574e-11eb-9321-cb785832ca7d.png)

Related issues:
https://github.com/grafana/grafana/issues/26152
https://github.com/grafana/grafana/issues/3035
https://github.com/grafana/grafana/issues/29026
https://github.com/grafana/grafana/issues/29817: GraphNG: superfluous getConfig() calls during init

**Comment by leeoniya** [comment]
@dprokop if you're tagging related issues, you've missed a few 🤣. sorted by most comments:

https://github.com/grafana/grafana/issues?q=is%3Aopen+label%3Aarea%2Fpanel%2Fgraph+sort%3Acomments-desc

the vast majority of these can be knocked out by Graph NG. i might be able to help with this soon ;)

**Comment by dprokop** [comment]
Was just tagging few that I have encountered today when going through the platform backlog. Thanks @leeoniya

**Comment by dprokop** [comment]
Lol, a few 🤣🤣🤣

**Comment by torkelo** [comment]
Still think the old graph grid colors & spacings (tick distances) are better, and the time axis being single line instead of 2 lines makes the graph height much more efficient, think we should get rid of the unconventional x-axis time format where the first x-axis tick is two lines and different from the others. It's interesting but the space it needs makes the whole graph look worse. 
![Screenshot from 2020-11-06 09-34-02](https://user-images.githubusercontent.com/10999/98344271-4dffa880-2013-11eb-8c0e-527ee8d62dc3.png)

**Comment by torkelo** [comment]
when I switch to the new graph panel it's always blank first render (in edit mode), have to trigger refresh for it to show (I see only legend, not the graph)

**Comment by torkelo** [comment]
Digging into the code more, I think we have a big problem with all these hooks and geometries. So many in-efficiencies and function call that update a mutable state for every series or axis. This seems insanely in-efficient. The way the config object is composed through potentially hundreds of components that all call a big expensive config hook and state one by one triggers a huge useEffect execution queue. All this could be done sync in one process. 

We also have some problem with first render triggers 4 renders, (one call to initPlot and 3 setData).

**Comment by torkelo** [comment]
@leeoniya is there a way to align grid lines and ticks with the edges of the graph? Looks a lot nicer when graph is framed by grid lines 

Show uplot at the top and flot at the bottom:

<img width="833" alt="Screen Shot 2020-11-08 at 09 04 13" src="https://user-images.githubusercontent.com/10999/98460092-75847b80-21a1-11eb-8afa-7edb71655d08.png">

**Comment by torkelo** [comment]
Updated description todo, and added uPlot issue https://github.com/leeoniya/uPlot/issues/362

**Comment by torkelo** [comment]
Fixed the x-axis formats so each tick is the same  (no rollover), 

![Screenshot from 2020-11-08 16-56-26](https://user-images.githubusercontent.com/10999/98470035-61fa0480-21e3-11eb-8874-aec699f46b55.png)

But failed to replicate our existing dynamic where we have move control over number of x-axis ticks and format can be based on time range:
https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Graph/utils.ts#L114

Some times uPlot generates to many ticks and they become overlapping 
![Screenshot from 2020-11-08 16-07-18](https://user-images.githubusercontent.com/10999/98468824-82729080-21dc-11eb-94b2-2b576f191867.png)

**Comment by torkelo** [comment]
yay, found the space property! it fixes the overlapping time labels
![Screenshot from 2020-11-08 17-15-28](https://user-images.githubusercontent.com/10999/98470463-11d07180-21e6-11eb-891f-490b2ece06cb.png)

**Comment by simPod** [comment]
What is the proper way to test uplot graphs when I'm on master? There's no option in UI it seems, should I somehow go through json schema?

**Comment by rawkode** [comment]
I posted a tweet today about a feature I'm missing from InfluxDB 2 since moving to Grafana. InfluxDB makes the active series prominent during hover.

I've got the raw gif on my laptop, I'll swap out the tweet link when off mobile 😃

https://twitter.com/rawkode/status/1340594353253863426?s=19

**Comment by leeoniya** [comment]
the underlyimg capability will be there: https://leeoniya.github.io/uPlot/demos/focus-cursor.html

i'm assuming it'll be exposed somehow as part of the legend/tooltip work being done in https://github.com/grafana/grafana/pull/29575

**Comment by dprokop** [comment]
@simPod the uPlot graph panel is an alpha one currently. If you want to test it on the master you need enable alpha panels in your config file:

```
[panels]
enable_alpha = true
```

**Comment by hugohaggmark** [comment]
When adding support for adding annotations, we would need to include the addition to Angular Graph in #29990

**Comment by leeoniya** [comment]
@dprokop i don't think show/hide grid is actually done, but it's checked off? it's a popular request (https://github.com/grafana/grafana/issues/9305), but from what i can tell it's not addressed yet.

**Comment by dprokop** [comment]
Closing in favour of #30564
